fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexBuildState.java

```diff
@@ -0,0 +1,160 @@
+ * </ul>
+ * @see #getIndexBuildStateAsync(FDBRecordStore, Index)
+ */
+@API(API.Status.UNSTABLE)
+public class IndexBuildState {
+    @Nonnull
+    private final IndexState indexState;
+    @Nullable
+    private final Long recordsScanned;
+    @Nullable
+    private final Long recordsInTotal;
+
+    /**
+     * Get the build progress ({@link IndexBuildState}) of the given index in the given record store asynchronously.
+     * @param store the record store containing the index
+     * @param index the index needed to be checked
+     * @return a future that completes to the index build state
+     */
+    public static CompletableFuture<IndexBuildState> getIndexBuildStateAsync(FDBRecordStore store, Index index) {
+        IndexState indexState = store.getIndexState(index);
+        if (indexState != IndexState.WRITE_ONLY) {
+            return CompletableFuture.completedFuture(new IndexBuildState(indexState));
+        }
+        return getRecordsScannedAsync(store, index).thenCompose(scannedRecords -> {
+            try {
+                return store.getSnapshotRecordCount()
+                        .thenApply(totalRecords -> new IndexBuildState(indexState, scannedRecords, totalRecords));
+            } catch (Exception ex) {
+                // getSnapshotRecordCount failed, very likely it is because there is no suitable COUNT type index
+                // defined.
+                return CompletableFuture.completedFuture(new IndexBuildState(indexState, scannedRecords, null));
+            }
+        });
+    }
+
+    /**
+     * Get the number of records successfully scanned and processed during the online index build process
+     * asynchronously.
+     * <p>
+     * This does not include the records scanned in the {@link OnlineIndexer} whose
+     * {@link OnlineIndexer.Builder#setTrackProgress(boolean)} is set to false.
+     * </p>
+     * @param store the record store containing the index
+     * @param index the index needed to be checked
+     * @return a future that completes to the total records scanned.
+     */
+    public static CompletableFuture<Long> getRecordsScannedAsync(FDBRecordStore store, Index index) {
+        return store.getContext().ensureActive()
+                .get(OnlineIndexer.indexBuildScannedRecordsSubspace(store, index).getKey())
+                .thenApply(FDBRecordStore::decodeRecordCount);
+    }
+
+    /**
+     * Get the index state.
+     * @return the index state
+     */
+    public IndexState getIndexState() {
+        return indexState;
+    }
+
+    /**
+     * Get the number of records successfully scanned and processed during the online index build process.
+     * <p>
+     * This does not include the records scanned in the {@link OnlineIndexer} whose
+     * {@link OnlineIndexer.Builder#setTrackProgress(boolean)} is set to false.
```

Comment:
Maybe something like, "If the `OnlineIndexer` has been configured to disable tracking progress (by setting `setTrackProgress` to `false`), then the number returned will not reflect the actual number of records scanned." Or something like that?

I wonder if it is worth also clarifying here that this won't include all records added during the build, or if that should be only in one place (at the top, probably) and the other place(s) we want to mention it should reference that one place.

Reply:
Fixed wordings for "track progress".

I didn't include clarifications for "all records added during the build" here because strictly speaking they are not "records scanned". But this clarification is useful when it is put under the context of comparing with the records in total. So I guess the top of `IndexBuildState` might be a better place. Added a point to the top.